source/uwp/Renderer/lib/AdaptiveActionRendererRegistration.cpp

```diff
@@ -0,0 +1,58 @@
+#include "pch.h"
+#include "AdaptiveActionRendererRegistration.h"
+#include "Util.h"
+
+using namespace Microsoft::WRL;
+using namespace ABI::AdaptiveCards::XamlCardRenderer;
+using namespace ABI::Windows::UI;
+
+namespace AdaptiveCards { namespace XamlCardRenderer
+{
+    AdaptiveActionRendererRegistration::AdaptiveActionRendererRegistration()
+    {
+    }
+
+    HRESULT AdaptiveActionRendererRegistration::RuntimeClassInitialize() noexcept try
+    {
+        m_registration = std::make_shared<RegistrationMap>();
+        return S_OK;
+    } CATCH_RETURN;
+
+    _Use_decl_annotations_
+    HRESULT AdaptiveActionRendererRegistration::Set(HSTRING type, IAdaptiveActionRenderer* renderer)
+    {
+        ComPtr<IAdaptiveActionRenderer> localRenderer(renderer);
+        (*m_registration)[TypeAsRegistrationKey(type)] = localRenderer;
+
+        return S_OK;
+    }
+
+    _Use_decl_annotations_
+    HRESULT AdaptiveActionRendererRegistration::Get(HSTRING type, IAdaptiveActionRenderer** result)
+    {
+        *result = nullptr;
+
+        RegistrationMap::iterator found = m_registration->find(TypeAsRegistrationKey(type));
+        if (found != m_registration->end())
+        {
+            *result = found->second.Get();
+        }
+        return S_OK;
+    }
+
+    _Use_decl_annotations_
+    HRESULT AdaptiveActionRendererRegistration::Remove(_In_ HSTRING type)
+    {
+        m_registration->erase(TypeAsRegistrationKey(type));
+        return S_OK;
+    }
+
+    _Use_decl_annotations_
+    std::string AdaptiveActionRendererRegistration::TypeAsRegistrationKey(_In_ HSTRING type)
```

Comment:
>TypeAsRegistrationKey [](start = 52, length = 21)

We've got this function in three places now, maybe make it a utility?

Reply:
I had it as an util and for some reason decided to make it a class member. I can't remember if I had a good reason for it, so that means I probably didn't.